### ISSUE TITLE
DisplayImage.ImageComponent.paint - ImageObserver not called

### ISSUE THREAD

**Issue by jonhare** [openimaj/openimaj#38]
OP-39: Scenario:
OpenIMAJ is integrated into a Swing Image/WebCam editing app. Video displays in JPanel, the client app wants to hook up for ImageObserver notifications, but ImageObserver is never called.

Details:
ImageComponent defines image observer with SwingUtilities.getRoot(this), which in my case is the JFrame. So I extend imageUpdate with a simple System.out.println("*") but statement is never printed.

Further analysis:
To eliminate threading issues, I modified ImageComponent to make its own instance an ImageObserver. I extended imageUpdate(java.awt.Image...) added same System.out.println("*"), modified g.drawImage(img, 0, 0, drawWidth, drawHeight, this) and once again, no sign of observer being called.

Unrelated but nice to have:
Would be nice to extend ImageComponent and define my own observer. In my "forked" version, rather than defining observer within paint, I defined two additional methods:

protected ImageObserver getDisplayImageObserver() {
 return SwingUtilities.getRoot(this);
}
protected ImageObserver getGridObserver() {
 return SwingUtilities.getRoot(this);
}

While this does not modify original implementation behavior, it give me flexibility to override either method and control which observer is hooked up to which image.

Kind Regards
Adam Zimowski

**Comment by jonhare** [comment]
Original reporter: @

**Comment by jonhare** [comment]
:  My apology for wrong assignment. This is not a Terrier issue, obviously.

**Comment by jonhare** [comment]
jonhare:  Dave's probably the best person to deal with Swing issues

**Comment by jonhare** [comment]
jonhare:  No problem with the wrong assignment - I've moved it to OpenIMAJ

**Comment by jonhare** [comment]
davedupplaw:  Hi Adam,

Just wondering - what does the client app need the ImageObserver for? 

You're right that it's set to SwingUtilities.getRoot(this) in the ImageComponent, but actually it's quite valid to be null here because the image being displayed is already fully in memory. I'm not entirely sure of the provenance of that little bit of code :)  The ImageObserver is mainly used for components that are waiting for the image to be delivered (say from a network) before it can be displayed; in which case drawImage() returns false and later on, when the image is available, the ImageObserver gets called to notify the classes waiting on the image that it is now available. In Swing it's often implemented using 'this' as all Components implement the ImageObserver interface and will repaint when the image becomes available.

But, as I said, the OpenIMAJ images are always available in memory and so no image observer is needed. You can see this explicitly as the drawImage() call in ImageComponent always returns true and this is also why the ImageObserver never gets called.

So, just wondering what you want your client to do? If you want it to be notified each time a new image is available then use the VideoDisplayListener interface with the VideoDisplay.addListener() method.

Hope that helps,
Dave

**Comment by jonhare** [comment]
:  Thanks for the quick reply. I thought that since image is preloaded by the grabber (this test case refers to usage of ImageComponent as part of cam live stream) this may be why observer is never called, but more generally, should ImageComponent simply pass null for ImageObserver since it won't get called? Under what scenario would overriding imageUpdate in root yield a working observer call?

Ultimately, what I am trying to do is get pixel under mouse from live frame display. Granted, showPixelColours=true natively displays it but it's in relation to translated source image, not the actual image displayed. In my case, ImageComponent lives inside JPanel (add, BorderLayout.CENTER) so frame could be stretched any way user resizes the window, and pixel under mouse does not map to neither to image.getRGB(mouseX, mouseY) nor ImageComponent.image.getRGB(pixelX, pixelY) because the image on display is never stored anywhere, it's rendered directly via g.drawImage(img, 0, 0, drawWidth, drawHeight, imgObserver). Note I'm not scaling image so my scaleFactor is not set.

This is apparent in the slightly modified version of your VideoCapture example. Notice, that pixel position is reported is not what really user would expect. If I point at left upper corner expecting 0,0 or something similar it is clear position is not reported off of actually displayed frame. Again, I can easily get position under mouse, but then, getting a corresponding pixel color is near impossible since the actual, stretched frame on display is not stored anywhere.

{code:title=VideoCaptureExample1.java}
package org.openimaj.examples.video.capture;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;

import javax.imageio.ImageIO;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JPanel;

import org.openimaj.image.DisplayUtilities.ImageComponent;
import org.openimaj.image.ImageUtilities;
import org.openimaj.image.MBFImage;
import org.openimaj.video.VideoDisplay;
import org.openimaj.video.VideoDisplayListener;
import org.openimaj.video.capture.VideoCapture;
import org.openimaj.video.capture.VideoCaptureException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Example showing how to open the default video capture device and display it
 * in a window. Also shows how to hook up a key listener so that the capture
 * pauses/unpauses when the spacebar is pressed.
 * 
 * @author Jonathon Hare (jsh2@ecs.soton.ac.uk)
 */
public class VideoCaptureExample1 extends KeyAdapter {
	
	private static final Logger log = LoggerFactory.getLogger(VideoCaptureExample1.class);
	VideoCapture vc;
	VideoDisplay<MBFImage> display;
	Thread displayThread;

	VideoCaptureExample1() throws VideoCaptureException {
		// open the capture device and create a window to display in 320, 240
		vc = new VideoCapture(1024, 768);

		
		final ImageComponent ic = new ImageComponent();
		ic.setAllowZoom( false );
		ic.setAllowPanning( false );
		ic.setTransparencyGrid( false );
		ic.setShowPixelColours( true );
		ic.setShowXYPosition( true );		

				JFrame win = new JFrame("Whooooooooo");
				win.setPreferredSize(new Dimension(320, 300));
				JPanel panel = new JPanel(new BorderLayout());
				JPanel buttonPanel = new JPanel();
				JButton save = new JButton("Save");
				JButton stop = new JButton("Stop");
				save.addActionListener(new ActionListener() {
					int count = 0;
		
					@Override
					public void actionPerformed(ActionEvent arg0) {
						MBFImage img = display.getVideo().getCurrentFrame();
						BufferedImage buff = ImageUtilities
								.createBufferedImageForDisplay(img);

						try {
							ImageIO.write(buff, "jpg", new File("/home/zima/cyc"
									+ count++ + ".jpg"));
						} catch (IOException e) {
							e.printStackTrace();
						}		
					}
				});
				stop.addActionListener(new ActionListener() {
					@Override
					public void actionPerformed(ActionEvent arg0) {
						log.debug("stopping..");
						vc.stopCapture();
						vc.close();
						displayThread = null;
					}
				});
				buttonPanel.add(save);
				buttonPanel.add(stop);
				panel.add(ic, BorderLayout.CENTER);
				panel.add(buttonPanel, BorderLayout.SOUTH);
				win.getContentPane().add(panel);
				win.pack();
				win.setVisible(true);
				win.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
				display = new VideoDisplay<MBFImage>( vc, null, ic );
				
				display.addVideoListener(new VideoDisplayListener< MBFImage >(){
					
					MBFImage lastImage = null;
					
					@Override
					public void afterUpdate(VideoDisplay<MBFImage> display) {
					}

					@Override
					public void beforeUpdate(MBFImage frame) {
						/*
						// face detection
						FaceDetector<DetectedFace, FImage> fd = new HaarCascadeDetector(20);
						List<DetectedFace> faces = fd.detectFaces(Transforms.calculateIntensity(frame));
						for(DetectedFace face : faces) {
							frame.drawShape(face.getBounds(), RGBColour.BLUE);
						}
						*/
						/*
						if(lastImage != null) {
							MBFImage tmp = frame.clone();
							frame.subtractInplace(lastImage).abs();
							lastImage = tmp;
						}
						else {
							lastImage = frame.clone();
						}
						*/
						//frame.processInplace(new CannyEdgeDetector2());
					}
				});
				
				displayThread = new Thread(display);
				displayThread.start();
	}

	@Override
	public void keyPressed(KeyEvent e) {
		// space pauses the video
		if (e.getKeyChar() == ' ')
			display.togglePause();
	}

	/**
	 * Main method.
	 * 
	 * @param args
	 *            ignored
	 * @throws VideoCaptureException
	 *             if their is a problem with the video capture hardware
	 */
	public static void main(String[] args) throws VideoCaptureException {
		new VideoCaptureExample1();
	}
}
{code}

Adam

**Comment by jonhare** [comment]
:  Here is a complete example file.

**Comment by jonhare** [comment]
davedupplaw:  Thanks for the sample file. You're right, the coordinates and colours that are shown are incorrect when the image is stretched by the component resizing. This is a bug in the ImageComponent which I'll try to fix. I'll also add a getRGBAtMouse() method (or something like that) so you can get the pixel colour at the mouse more easily.

**Comment by jonhare** [comment]
:  Thank much. I've fixed this in my forked version by double buffering the frame display image. Translation at paint time seems like a bad idea as it drastically degrates performance. I'm very much looking forward to your solution. Please share, if you don't mind.

**Comment by jonhare** [comment]
davedupplaw:  I've just committed a new revision of the DisplayUtilities with a few changes to the ImageComponent. In particular, it now correctly reports the pixel colour at the mouse position whatever scale the image is at (even when a different scale is used in the horizontal or vertical dimensions). The component will detect resizes and set the scale factor accordingly when autoFit is TRUE (using setAutoFit(boolean)). You can choose to keep the aspect ratio the same or not (using setKeepAspect(boolean)). There are now accessors for you to get information about where the mouse is and over what it is hovering.  getCurrentMousePosition() returns the mouse position in the component in pixels. getCurrentMouseImagePosition() returns the pixel coordinate of the displayed image that is under the mouse.  getCurrentPixelColour() will return the colour of that pixel under the mouse. If you haven't used setOriginalImage() this method will return the BufferedImage pixel values (RGB, 0-255); if you set the OpenIMAJ image object using setOriginalImage() it will return the pixel values from the OpenIMAJ image object.

I hope these updates allow you to do what you want to do. Let us know if they don't.

I'm going to close this as actually the reported bug isn't really a bug at all (ImageObservers shouldn't be called by the ImageComponent as the displayed images are in memory).

Dave

**Comment by jonhare** [comment]
davedupplaw:  
RESOLVED: Won't fix.

ImageObservers shouldn't get called as the images being displayed are fully in memory. Therefore there is nothing to fix.

However, I made the code pass null as the ImageObserver in the drawImage() methods as there was some spurious and confusing code in this part.

The discussion highlighted a few other bugs which have been subsequentally fixed.

**Comment by jonhare** [comment]
:  Dave, thanks for the code update. Unfortunately, this seems to introduce two problems: 1) Major performance delay when streaming video as compared to previous version 2) NPE thrown from MouseMove when mouse is moved over the viewport but frame hasn't rendered yet.

I am reverting to old version for now. I do have it working if you care to see my code, though I only work from the video streaming perspective, haven't done much with image module. In any case, you can see the performance by running the VideoCaptureFramesExample I added in OP-42 with old revision of DisplayUtilities and the new one.

Kind Regards,
Adam

PS. The API (getCurrentPixel and friends) is very straighforward and was very flexible for my needs.
